File: cpp/longest_substring_without_repeating_characters.cpp
/**
 * @file longest_substring_without_repeating_characters.cpp
 * @brief https://leetcode.com/problems/longest-substring-without-repeating-characters/
 * @date 2025-03-11
 */
#include <cassert>
#include <string>
#include <unordered_map>
using namespace std;

int lengthOfLongestSubstring(string s) {
	size_t max_count = 0;
	std::unordered_map<char, size_t> char_to_index;
	for (size_t i = 0; i < s.size(); ++i) {
		if (char_to_index.contains(s[i])) {
			i = char_to_index[s[i]];
			char_to_index.clear();
		} else {
            char_to_index[s[i]] = i;
            max_count = std::max(char_to_index.size(), max_count);
        }
	}
	return static_cast<int>(max_count);
}

int main() {
	assert(lengthOfLongestSubstring("tmmzuxt") == 5);
	assert(lengthOfLongestSubstring("abcabcbb") == 3);
	return 0;
}

File: python/longest_substring_without_repeating_characters.py
"""
https://leetcode.com/problems/longest-substring-without-repeating-characters/
"""


def lengthOfLongestSubstring(s: str) -> int:
    left = 0
    count: set[str] = set()
    max_length = 0
    for right in range(len(s)):
        if s[right] in count:
            while True:
                count.remove(s[left])
                left += 1
                if s[right] == s[left - 1]:
                    break
        count.add(s[right])
        max_length = max(max_length, right - left + 1)
    return max_length


if __name__ == '__main__':
    assert lengthOfLongestSubstring("tmmzuxt") == 5
    assert lengthOfLongestSubstring("abcabcbb") == 3
